### PR TITLE
Add the ability to toggle auto scale down in data_science stack

### PR DESCRIPTION
**Pull request by kenoir**
### What is this PR trying to achieve?

Allow @harrisonpim to keep instances running overnight for long running processes.

### Who is this change for?

Science 🔬 !

### CODE REVIEW

**Comment by kenoir** [comment]
Resolves: https://github.com/wellcometrust/platform/pull/2447

**Comment by kenoir** [comment]
Resolves https://github.com/wellcometrust/platform/issues/2435

File: data_science/scripts/toggle_asg.py [hunk 1]
```diff
@@ -42,6 +81,14 @@
             desired_size=desired_size
         )
 
+    # Scaling rules
+    elif args['--keep-on'] or args['--allow-scale-down']:
+        if args['--keep-on']:
+            keep_on(asg_client, asg_name)
+        else:
+            allow_scale_down(asg_client, asg_name)
+
```
Comment (mklander) [inline data_science/scripts/toggle_asg.py]: Can these be a single switch and switch between DesiredCapacity 1 or 0?  It's a bit confusing to allow both `keep-on` and `allow-scale-down` on the command line.

File: data_science/scripts/toggle_asg.py [hunk 2]
```diff
@@ -3,63 +3,145 @@
 """
 Start/stop all the instances in an autoscaling group.
 
-Usage: toggle_asg.py (--start | --stop | --status) [--type=(p2 | t2)]
+Usage: toggle_asg.py (--start | --stop | --enable-overnight | --disable-overnight) [--type=(p2 | t2)]
+       toggle_asg.py [--type=(p2 | t2)]
```
Comment (alexwlchan) [inline data_science/scripts/toggle_asg.py]: Consider adding `toggle_asg.py -h | --help` to this list, so you can get a usage message.

Currently it always does something, so getting to this message is quite hard!

Also, maybe `--start` et al should be non-prefix flags? So more like:

```
Usage: toggle_asg.py (start | stop | enable-overnight | disable-overnight | status) [--type=(p2 | t2)]
       toggle_asg.py -h | --help
```

File: data_science/scripts/toggle_asg.py [hunk 3]
```diff
@@ -3,63 +3,145 @@
...
+
 
 Actions:
   --start                 Start the autoscaling group (set the desired count to 1).
   --stop                  Stop the autoscaling group (set the desired count to 0).
-  --status                Report the current side of autoscaling group.
+  --enable-overnight      Keeps the autoscaling group desired count at 1 (disables auto scale down)
```
Comment (alexwlchan) [inline data_science/scripts/toggle_asg.py]: Describe in terms of user visible effects: “Keeps the ASG running overnight (disables auto scale down)” and “Stops the ASG from running overnight (enables auto scale down)”

File: data_science/scripts/toggle_asg.py [hunk 4]
```diff
@@ -3,63 +3,145 @@
...
+    )
+
+
+def get_status(client, tag_name, scheduled_action_name):
+    """Gets the status of an autoscaling group.
+
+    Keyword arguments:
+    client -- boto3 autoscaling client
+    tag_name -- tags used to identify autoscaling group
+    scheduled_action_name -- name of the scheduled action to get status for
```
Comment (alexwlchan) [inline data_science/scripts/toggle_asg.py]: 👍

File: data_science/scripts/toggle_asg.py [hunk 5]
```diff
@@ -3,63 +3,145 @@
...
+    desired_capacity = scaling_group['DesiredCapacity']
+    instance_count = len(scaling_group['Instances'])
+    scaling_group_name = scaling_group['AutoScalingGroupName']
+
+    response = client.describe_scheduled_actions(
+        AutoScalingGroupName=scaling_group_name,
+        ScheduledActionNames=[scheduled_action_name]
+    )
+
+    scheduled_desired_capacity = response['ScheduledUpdateGroupActions'][0]['DesiredCapacity']
```
Comment (alexwlchan) [inline data_science/scripts/toggle_asg.py]: What is ScheduledUpdateGroupActions doesn't have length 1? Maybe add an assert that prints it if not, so we get some debugging info if something goes wrong.

**Pull request merged**